packages/dd-trace/src/config.js

```diff
@@ -10,19 +11,88 @@ const { isTrue, isFalse } = require('./util')
 
 const runtimeId = `${id().toString()}${id().toString()}`
 
-class Config {
+function pluginEnv (name) {
+  return platform.env(`DD_TRACE_${name.toUpperCase()}`.replace(/[^a-z0-9_]/ig, '_'))
+}
+
+class Config extends EventEmitter {
   constructor (options) {
+    super()
+    this._hasEmitted = {}
+    this.configure(options)
+  }
+
+  reset () {
+    for (const prop of Reflect.ownKeys(this)) {
+      delete this[prop]
+    }
+    EventEmitter.call(this)
+    this._hasEmitted = {}
+    this.configure()
+  }
+
+  configurePlugin (name, config = {}) {
+    this.pluginConfigs = this.pluginConfigs || {}
+
+    if (!name) {
+      return
+    }
+
+    if (typeof config === 'boolean') {
+      config = { enabled: config }
+    }
+
+    const enabled = pluginEnv(`${name}_ENABLED`)
+    if (enabled !== undefined) {
+      config.enabled = isTrue(enabled)
+    }
+
+    const analyticsEnabled = pluginEnv(`${name}_ANALYTICS_ENABLED`)
+    const analyticsSampleRate = Math.min(Math.max(pluginEnv(`${name}_ANALYTICS_SAMPLE_RATE`), 0), 1)
+
+    if (isFalse(analyticsEnabled)) {
+      config.analytics = false
+    } else if (!Number.isNaN(analyticsSampleRate)) {
+      config.analytics = analyticsSampleRate
+    } else if (isTrue(analyticsEnabled)) {
+      config.analytics = true
+    }
+
+    this.pluginConfigs[name] = Object.assign(this.pluginConfigs[name] || {}, config)
+
+    this.retroEmit(`update.plugin.${name}`) // TODO this is not needed while the instrumenter managers plugin state
+    this.retroEmit('update.plugins')
+  }
+
+  retroOn (name, handler) {
```

Comment:
I had initially taken that approach, but I decided I wasn't fond of the pattern sometimes used in Node.js core where `on` handlers can run immediately/synchronously if the event has already fired. `retroOn` (i.e. "retrospectively, on") makes it clearer what's actually happening. It isn't an implementation detail, because consumers of this class need to be able to reason about when their code is actually going to be run, and there's a huge different between "maybe later, maybe right now" and "definitely later".